source/include/ble/cs_CrownstoneCentral.h

```diff
@@ -0,0 +1,200 @@
+	 *
+	 * @return ERR_WRONG_STATE         When already connected.
+	 * @return ERR_BUSY                An operation is in progress (connect, write, disconnect).
+	 * @return ERR_WAIT_FOR_SUCCESS    When the connection will be attempted. Wait for EVT_CS_CENTRAL_CONNECT_RESULT.
+	 */
+	cs_ret_code_t connect(stone_id_t stoneId, uint16_t timeoutMs = 3000);
+
+	/**
+	 * Connect, discover, and read session data.
+	 *
+	 * @param[in] address              The MAC address of the crownstone you want to connect to.
+	 * @param[in] timeoutMs            How long (in milliseconds) to try connecting, before giving up.
+	 *
+	 * @return ERR_WRONG_STATE         When already connected.
+	 * @return ERR_BUSY                An operation is in progress (connect, write, disconnect).
+	 * @return ERR_WAIT_FOR_SUCCESS    When the connection will be attempted. Wait for EVT_CS_CENTRAL_CONNECT_RESULT.
+	 */
+	cs_ret_code_t connect(const device_address_t& address, uint16_t timeoutMs = 3000);
+
+	/**
+	 * Terminate current connection.
+	 *
+	 * @return ERR_SUCCESS             When already disconnected.
+	 * @return ERR_WAIT_FOR_SUCCESS    When disconnecting. Wait for EVT_BLE_CENTRAL_DISCONNECTED.
+	 */
+	cs_ret_code_t disconnect();
+
+	/**
+	 * Write a control command, and get the result.
+	 *
+	 * When the result code in the result data is ERR_WAIT_FOR_SUCCESS, you will get another EVT_CS_CENTRAL_WRITE_RESULT event.
+	 *
+	 * TODO: add timeout.
+	 *
+	 * @param[in] commandType          What control command.
+	 * @param[in] data                 Pointer to the command payload data.
+	 * @param[in] size                 Size of the command payload data.
+	 *
+	 * @return ERR_WAIT_FOR_SUCCESS    When the write is started. Wait for EVT_CS_CENTRAL_WRITE_RESULT.
+	 */
+	cs_ret_code_t write(cs_control_cmd_t commandType, uint8_t* data, uint16_t size);
+
+	/**
+	 * Request the write buffer, to avoid having to copy data to the write buffer.
+	 *
+	 * @return     When busy:          A null pointer.
+	 * @return     On success:         A pointer to the write buffer, and the length of the buffer.
+	 */
+	cs_data_t requestWriteBuffer();
+
+private:
+	enum ServiceIndex {
+		SERVICE_INDEX_CROWNSTONE = 0,
+		SERVICE_INDEX_SETUP = 1,
+		SERVICE_INDEX_DEVICE_INFO = 2,
+		SERVICE_INDEX_DFU = 3,
+	};
+
+	enum class Operation: uint8_t {
+		NONE,
+		CONNECT,
+		WRITE
+	};
+
+	enum class ConnectSteps: uint8_t {
+		NONE = 0,
+		GET_ADDRESS,
+		CONNECT,
+		DISCOVER,
+		ENABLE_NOTIFICATIONS,
+		SESSION_KEY,
+		SESSION_DATA,
+		DONE
+	};
+
+	enum class WriteControlSteps: uint8_t {
+		NONE = 0,
+		WRITE,
+		RECEIVE_RESULT,
+		DONE
+	};
+
+	UUID _uuids[4];
+
+	uint16_t _sessionKeyHandle;
+	uint16_t _sessionDataHandle;
+	uint16_t _controlHandle;
+	uint16_t _resultHandle;
+	uint16_t _resultCccdHandle;
+	OperationMode _opMode = OperationMode::OPERATION_MODE_UNINITIALIZED;
+	Operation _currentOperation = Operation::NONE;
+	uint8_t _currentStep;
```

Comment:
If _currentStep can be of type ConnectSteps or WriteControlSteps have you considered:
- putting the enum values together in one type, or
- using union {ConnectSteps a; WriteControlSteps} as type for _currentStep, or
- two separate fields _currentWriteStep and _currentConnectStep?

Reply:
I added some explanation for all these variables.

A union could be a nice solution, but would it really improve the code a lot?
There can only be 1 current step, so i prefer having only 1 variable.